src/tallies/derivative.cpp

```diff
@@ -0,0 +1,719 @@
+#include "openmc/error.h"
+#include "openmc/material.h"
+#include "openmc/nuclide.h"
+#include "openmc/settings.h"
+#include "openmc/tallies/tally.h"
+#include "openmc/tallies/derivative.h"
+#include "openmc/xml_interface.h"
+
+#include <sstream>
+
+template class std::vector<openmc::TallyDerivative>;
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+namespace model {
+  std::vector<TallyDerivative> tally_derivs;
+  std::unordered_map<int, int> tally_deriv_map;
+}
+
+//==============================================================================
+// TallyDerivative implementation
+//==============================================================================
+
+TallyDerivative::TallyDerivative(pugi::xml_node node)
+{
+  if (check_for_node(node, "id")) {
+    id = std::stoi(get_node_value(node, "id"));
+  } else {
+    fatal_error("Must specify an ID for <derivative> elements in the tally "
+                "XML file");
+  }
+
+  if (id <= 0)
+    fatal_error("<derivative> IDs must be an integer greater than zero");
+
+  std::string variable_str = get_node_value(node, "variable");
+
+  if (variable_str == "density") {
+    variable = DIFF_DENSITY;
+
+  } else if (variable_str == "nuclide_density") {
+    variable = DIFF_NUCLIDE_DENSITY;
+
+    std::string nuclide_name = get_node_value(node, "nuclide");
+    bool found = false;
+    for (auto i = 0; i < data::nuclides.size(); ++i) {
+      if (data::nuclides[i]->name_ == nuclide_name) {
+        found = true;
+        //TODO: off-by-one
+        diff_nuclide = i + 1;
+      }
+    }
+    if (!found) {
+      std::stringstream out;
+      out << "Could not find the nuclide \"" << nuclide_name
+          << "\" specified in derivative " << id << " in any material.";
+      fatal_error(out);
+    }
+
+  } else if (variable_str == "temperature") {
+    variable = DIFF_TEMPERATURE;
+
+  } else {
+    std::stringstream out;
+    out << "Unrecognized variable \"" << variable_str
+        << "\" on derivative " << id;
+    fatal_error(out);
+  }
+
+  diff_material = std::stoi(get_node_value(node, "material"));
+}
+
+//==============================================================================
+// Non-method functions
+//==============================================================================
+
+extern "C" void
+read_tally_derivatives(pugi::xml_node* node)
+{
+  // Populate the derivatives array.  This must be done in parallel because
+  // the derivatives are threadprivate.
+  #pragma omp parallel
+  {
+    for (auto deriv_node : node->children("derivative"))
+      model::tally_derivs.push_back(deriv_node);
```

Comment:
It's not really implicit conversion right?  My understanding is that deriv_node is being passed to the explicit constructor of `TallyDerivative` and copied into the vector (except the copy is likely elided away here).  My general philosophy for `push_back` vs `emplace_back` was to always use `push_back` for stuff that can be safely copied.

Reply:
It is implicit conversion. When you push back to a `std::vector<T>` the value you are pushing back should be a `T`, otherwise it has to be implicitly converted. So, if you were to have
```C++
explicit TallyDerivative(pugi::xml_node node)
```
this wouldn't compile. Anyway, this type is not big or complicated enough that the copy would really matter either way, so I don't feel strongly.

Reply:
Ah, so it is!  Okay, I'll use, `push_back`, and I'll also declare that constructor explicit